cms_coding_rules.md

```diff
@@ -0,0 +1,144 @@
+# CMS Naming, Coding, And Style Rules
+
+### By the CMS Ofﬂine Software Development Team
+
+## Abstract
+This document describes the design, naming, coding, and style rules and recommendations for CMS software written in C++, plus a few guidelines for Python configuration.
+
+
+## Outline
+#### 1 -- Introduction
+#### 2 -- Naming Rules
+#### 3 -- Style Rules
+#### 4 -- Technical Coding Rules
+#### 5 -- Documentation Rules
+#### 6 -- Packaging Rules
+#### 7 -- Design and Coding Guidelines
+
+## 1 -- Introduction
+This document describes the CMS C++ software naming, coding, style, and documentation rules and recommendations.
+All CMS C++ software is expected to comply with the rules. The asterisk (\*) after some rules indicates that there may be exceptional use cases where the rule may be violated with good justiﬁcation.
+Coding rules are meant to prevent serious problems in software function, performance, maintainability, usability, and portability. The Packaging Rules section also has some brief guidelines for Python configuration.
+
+## 2 -- Naming Rules
+1. C++ header ﬁles use the sufﬁx .h, e.g. `CaloCluster.h`. (\*)
+2. For C++ source ﬁles, the preferred sufﬁx  is .cc, e.g. `CaloCluster.cc`, though there are existing files using .cpp and .cxx suffixes. (\*)
+3. For a header file that contains a class, name that ﬁle after the class.
+4. Name source ﬁles after the class.
+5. For class, struct, type, and enumeration names use upper class initials, e.g. `GeometryBuilder`.
+6. For namespaces use lower case, e.g. `namespace edm`.
+7. Start method names with lowercase, use upper case initials for following words, e.g. `collisionPoint()`.  
+  Allowed exception: Implementation of virtual methods inherited from external packages, e.g. `ProcessHits()` method required by Geant4.  
+8. Start data member names with lower case. A trailing "\_" is the preferred method to distinguish a data member from the getter method (e.g. `momentum_`).
+9. Using "set" for a setter method is preferred, e.g. `setMomentum(double m)`.
```

Comment:
according to 2.11, we should not use single character names. we should update the example above e.g `setMomentum(double value)`

Reply:
Thanks for spotting this. The rules were not even self-consistent! Fixed now.